src/helpers/ring.rs

```diff
@@ -26,183 +27,270 @@ pub enum HelperAddr {
+    /// into a ring.
+    fn ring_channel(&self, id: ProtocolId) -> Self::ChannelType;
+}
+
 /// Trait for MPC helpers to communicate with each other. Helpers can send messages and
 /// receive messages from a specific helper.
 #[async_trait]
 pub trait Ring {
     /// Send message to the destination. Implementations are free to choose whether it is required
     /// to wait until `dest` acknowledges message or simply put it to a outgoing queue
-    async fn send<T: Message>(&self, dest: HelperAddr, msg: T) -> Result<(), Error>;
-    async fn receive<T: Message>(&self, source: HelperAddr) -> Result<T, Error>;
+    async fn send<T: Message>(&mut self, dest: HelperAddr, msg: T) -> Result<(), Error>;
+
+    /// Receive a given message from the `source`
+    async fn receive<T: Message>(&mut self, source: HelperAddr) -> Result<T, Error>;
 }
 
 #[cfg(test)]
 pub mod mock {
     use crate::helpers::error::Error;
-    use crate::helpers::ring::{HelperAddr, Message, Ring};
+    use crate::helpers::error::Error::{ReceiveError, SendError};
+    use crate::helpers::ring::{CommunicationGateway, HelperAddr, Message, ProtocolId, Ring};
     use async_trait::async_trait;
-    use std::any::TypeId;
     use std::collections::hash_map::Entry;
     use std::collections::HashMap;
     use std::sync::{Arc, Mutex};
-    use tokio::sync::mpsc::{channel, Sender};
+    use tokio::sync::mpsc::{channel, Receiver, Sender};
 
-    /// Internally we represent all messages to be a sequence of bytes and store them inside
-    /// a hashmap where each element is addressable by message type id and destination (i.e. who
-    /// is the intended receiver of this message).
-    type MessageBuf = HashMap<(HelperAddr, TypeId), Box<[u8]>>;
+    /// Entry point for all communications for a given helper. Keeps track of all protocols
+    /// started by it and routes messages based on protocol identifiers they carry on with them.
+    #[derive(Debug)]
+    pub struct TestHelper {
+        left: Sender<MessageEnvelope>,
+        right: Sender<MessageEnvelope>,
+        routing_table: Arc<Mutex<RoutingTable>>,
+    }
+
+    /// Communication channel for test helpers. Holds references to channel senders to send messages
+    /// to the outside and owns the receivers (one per peer) to accept messages sent to it.
+    /// Owner control the lifetime of protocol communication links:
+    /// when this instance is dropped, link is removed.
+    #[derive(Debug)]
+    pub struct TestChannel {
+        protocol_id: ProtocolId,
+        table: Arc<Mutex<RoutingTable>>,
+        senders: HashMap<HelperAddr, Sender<MessageEnvelope>>,
+        receivers: HashMap<HelperAddr, Receiver<MessageEnvelope>>,
+    }
 
-    /// Each message is packed inside an envelope with some meta information about it.
     #[derive(Debug)]
     struct MessageEnvelope {
         source: HelperAddr,
-        type_id: TypeId,
+        protocol_id: ProtocolId,
```

Comment:
As I mentioned elsewhere regarding `source`, I don't think that you need to include the message type/protocol ID on every message.  It's probably OK to have a little redundancy in tests, but that is what it is.